File: apps/web/src/services/writing.ts
/**
 * 写作助手 API
 */
import { config, fetchWithFallback } from '../config'

export interface WritingResult {
  topic: string
  title: string
  content: string
  used_words: string[]
  is_mock: boolean
}

export async function generateWriting(
  topic: string,
  words: string[] = [],
  style: string = 'academic',
  length: number = 150,
  instruction: string = '',
  taskLabel: string = '',
): Promise<WritingResult> {
  const proxyUrl = '/api/ai/writing'
  const directUrl = config.aiGateway
    ? `${config.aiGateway}/api/ai/writing`
    : proxyUrl

  const resp = await fetchWithFallback(proxyUrl, directUrl, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      topic,
      words,
      style,
      length,
      instruction,
      task_label: taskLabel,
    }),
  })

  if (!resp.ok) {
    const body = await resp.text().catch(() => '')
    throw new Error(`写作生成失败 ${resp.status}: ${body.slice(0, 100)}`)
  }
  return resp.json()
}

/* ============ 官方写作真题库 ============ */

export interface WritingPrompt {
  id: string
  exam: string
  taskType: string
  taskLabel: string
  prompt: string
  instruction: string
  source: string
  sourceUrl: string
}

export interface WritingPromptResult {
  items: WritingPrompt[]
  total: number
  exam: string
  source: string
}

export async function fetchWritingPrompts(
  params: { limit?: number; taskType?: string; shuffle?: boolean } = {},
): Promise<WritingPromptResult> {
  const qs = new URLSearchParams()
  qs.set('limit', String(params.limit ?? 10))
  if (params.taskType) qs.set('task_type', params.taskType)
  if (params.shuffle) qs.set('shuffle', 'true')

  const path = `/api/writing/prompts?${qs.toString()}`
  const directUrl = config.aiGateway ? `${config.aiGateway}${path}` : path

  const resp = await fetchWithFallback(path, directUrl, { method: 'GET' })
  if (!resp.ok) throw new Error(`写作真题库加载失败 ${resp.status}`)
  return resp.json()
}

File: services/ai-gateway/app/schemas/writing.py
"""写作助手 Schema"""
from __future__ import annotations

from pydantic import BaseModel, Field


class WritingRequest(BaseModel):
    topic: str = Field(..., min_length=1, max_length=200, description="写作主题")
    words: list[str] = Field(default=[], description="要求用到的单词")
    style: str = Field(default="academic", description="写作风格：academic/essay/daily/business")
    length: int = Field(default=150, ge=50, le=800, description="目标字数")
    instruction: str = Field(default="", max_length=2000, description="官方真题的写作指令（有则按真题范文模式生成）")
    task_label: str = Field(default="", max_length=80, description="任务类型标签，如 Analyze an Issue")


class WritingResponse(BaseModel):
    topic: str
    title: str = ""
    content: str = ""
    used_words: list[str] = []  # 实际用到的词
    is_mock: bool = False


# ============ 写作真题库 ============

class WritingPrompt(BaseModel):
    id: str
    exam: str = 'GRE'
    taskType: str = 'issue'
    taskLabel: str = ''
    prompt: str
    instruction: str
    source: str = ''
    sourceUrl: str = ''


class WritingPromptResponse(BaseModel):
    items: list[WritingPrompt]
    total: int
    exam: str = 'GRE'
    source: str = ''


class WritingPromptStats(BaseModel):
    total: int
    byType: dict[str, int] = {}
    source: str = ''
